tests/admin_views/test_autocomplete_view.py

```diff
@@ -0,0 +1,231 @@
+import json
+
+from django.contrib import admin
+from django.contrib.admin import site
+from django.contrib.admin.tests import AdminSeleniumTestCase
+from django.contrib.admin.views.autocomplete import AutocompleteJsonView
+from django.contrib.auth.models import Permission, User
+from django.contrib.contenttypes.models import ContentType
+from django.http import Http404
+from django.test import RequestFactory, override_settings
+from django.urls import reverse, reverse_lazy
+
+from .admin import AnswerAdmin, QuestionAdmin
+from .models import Answer, Author, Authorship, Book, Question
+from .tests import AdminViewBasicTestCase
+
+PAGINATOR_SIZE = AutocompleteJsonView.paginate_by
+
+
+class AuthorAdmin(admin.ModelAdmin):
+    search_fields = ['id']
+
+
+class AuthorshipInline(admin.TabularInline):
+    model = Authorship
+    autocomplete_fields = ['author']
+
+
+class BookAdmin(admin.ModelAdmin):
+    inlines = [AuthorshipInline]
+
+
+site.register(Question, QuestionAdmin)
+site.register(Answer, AnswerAdmin)
+site.register(Author, AuthorAdmin)
+site.register(Book, BookAdmin)
+
+
+class AutocompleteJsonViewTests(AdminViewBasicTestCase):
+    as_view_args = {'model_admin': QuestionAdmin(Question, site)}
+    factory = RequestFactory()
+    url = reverse_lazy('admin:admin_views_question_autocomplete')
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.user = User.objects.create_user(
+            username='user', password='secret',
+            email='user@example.com', is_staff=True,
+        )
+        super().setUpTestData()
+
+    def test_success(self):
+        q = Question.objects.create(question='Is this a question?')
+        request = self.factory.get(self.url, {'term': 'is'})
+        request.user = self.superuser
+        response = AutocompleteJsonView.as_view(**self.as_view_args)(request)
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.content.decode('utf-8'))
+        self.assertEqual(data, {
+            'results': [{'id': str(q.pk), 'text': q.question}],
+            'pagination': {'more': False},
+        })
+
+    def test_must_be_logged_in(self):
+        response = self.client.get(self.url, {'term': ''})
+        self.assertEqual(response.status_code, 200)
+        self.client.logout()
+        response = self.client.get(self.url, {'term': ''})
+        self.assertEqual(response.status_code, 302)
+
+    def test_has_change_permission_required(self):
+        """
+        Users require the change permission for the related model to the
+        autocomplete view for it.
```

Comment:
The requirement for the `change` permission means the autocomplete doesn't work for a user with only the new `view` permission on the related model.

See https://code.djangoproject.com/ticket/29502

This should probably be weakened. (No extra permission is needed here if you're not using autocomplete.) 

@codingjoe: Your thoughts welcomed! 🙂

Reply:
I wasn't even aware of the view permission. This is fine.